exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/DrillFunctionRegistry.java

```diff
@@ -218,4 +302,141 @@ private void registerOperatorsWithoutInference(DrillOperatorTable operatorTable)
       }
     }
   }
+
+  /**
+   * Function registry holder. Stores function implementations by jar name, function name.
+   */
+  private class GenericRegistryHolder<T, U> {
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    private final AutoCloseableLock readLock = new AutoCloseableLock(readWriteLock.readLock());
+    private final AutoCloseableLock writeLock = new AutoCloseableLock(readWriteLock.writeLock());
+
```

Comment:
Not convinced the locking strategy works. The read, write and read/write locks are independent: locking one does not block the others. This means that a write lock will mutate data structures while a read lock attempts to read them: probably not what was intended.